src/Microsoft.AspNetCore.Routing/Internal/UriBuildingContext.cs

```diff
@@ -197,6 +203,38 @@ public override string ToString()
             return _uri.ToString();
         }
 
+        private void EncodeValue(string value)
+        {
+            EncodeValue(value, encodeSlashes: true);
+        }
+
+        private void EncodeValue(string value, bool encodeSlashes)
+        {
+            EncodeValue(value, startIndex: 0, characterCount: value.Length, encodeSlashes);
+        }
+
+        private void EncodeValue(string value, int startIndex, int characterCount, bool encodeSlashes)
+        {
+            // Just encode everything if its ok to encode slashes
+            if (!encodeSlashes && value.IndexOf('/', startIndex, characterCount) > 0)
+            {
+                var subValue = value.Substring(startIndex, characterCount);
+                var parts = subValue.Split('/');
```

Comment:
We can avoid allocating with the `Substring` and `Split` with a loop using `value.IndexOf('/', lastSlashIndex, remainingCharacterCount)` to find the next slash and then writing that set of characters using `_urlEncoder.Encode(Writer, value, lastSlashIndex + 1, nextSlashIndex);` + a find call to Encode with the trailing chars.

Reply:
Good idea 👍